jdk-annotations/astubx-generator-cli/src/main/java/com/uber/nullaway/jdkannotations/AstubxGeneratorCLI.java

```diff
@@ -0,0 +1,251 @@
+    // get parsed JSON file
+    Map<String, List<ClassJson>> parsed = parseJson(jsonDirPath);
+
+    // check if the astubx file directory exists
+    try {
+      Files.createDirectories(Paths.get(astubxDirPath));
+    } catch (IOException e) {
+      System.err.println("Failed to create directory: " + astubxDirPath + e.getMessage());
+    }
+    File outputFile = new File(astubxDirPath, "output.astubx");
+
+    ImmutableMap<String, String> importedAnnotations =
+        ImmutableMap.of(
+            "NonNull", "org.jspecify.annotations.NonNull",
+            "Nullable", "org.jspecify.annotations.Nullable");
+
+    Map<String, Set<String>> packageAnnotations = new HashMap<>(); // not used
+    Map<String, Set<String>> typeAnnotations = new HashMap<>();
+    Map<String, MethodAnnotationsRecord> methodRecords = new LinkedHashMap<>();
+    Set<String> nullMarkedClasses = new LinkedHashSet<>();
+    Map<String, Set<Integer>> nullableUpperBounds = new LinkedHashMap<>(); // not used
+
+    for (Map.Entry<String, List<ClassJson>> entry : parsed.entrySet()) {
+      // for each class
+      for (ClassJson clazz : entry.getValue()) {
+        // get fully qualified class name
+        String fullyQualifiedClassName = clazz.type();
+        if (fullyQualifiedClassName.indexOf('<') != -1) {
+          fullyQualifiedClassName =
+              fullyQualifiedClassName.substring(0, fullyQualifiedClassName.indexOf('<'));
+        }
+        // (Is this needed? NullAway checks NullMarkedness with fully qualified name)
+        if (clazz.nullMarked()) {
+          nullMarkedClasses.add(fullyQualifiedClassName);
+        }
+
+        // check upperbounds of type parameters
+        Set<Integer> upperBoundIndex = new LinkedHashSet<>();
+        for (int idx = 0; idx < clazz.typeParams().size(); idx++) {
+          TypeParam typeParam = clazz.typeParams().get(idx);
+          for (String bound : typeParam.bounds()) {
+            if (bound.matches("@Nullable .*")) {
+              upperBoundIndex.add(idx);
+            }
+          }
+          if (!upperBoundIndex.isEmpty()) {
+            nullableUpperBounds.put(fullyQualifiedClassName, upperBoundIndex);
+          }
+        }
+
+        // get methodRecords
+        getMethodRecords(clazz, fullyQualifiedClassName, methodRecords);
+      }
+    }
+
+    // write astubx file
+    try (DataOutputStream out = new DataOutputStream(new FileOutputStream(outputFile))) {
+      StubxWriter.write(
+          out,
+          importedAnnotations,
+          packageAnnotations,
+          typeAnnotations,
+          methodRecords,
+          nullMarkedClasses,
+          nullableUpperBounds);
+    } catch (IOException e) {
+      System.err.println("Error writing JSON file: " + outputFile.getAbsolutePath());
+      throw new RuntimeException(e);
+    }
+
+    // return result as modelData for testing
+    AstubxData modelData = new AstubxData(methodRecords, nullableUpperBounds, nullMarkedClasses);
```

Comment:
This is not a clean way to do things.  This function should just return the `AstubxData`, and a separate function should take care of writing the `.astubx` file.